### PR TITLE
Extract a DimensionsPanel component as a reusable component between Global Styles and Block Inspector

### PR DESCRIPTION
**Pull request by youknowriad**
Related to #37064 and #47348 and similar to #47356 (and follow-ups)

## What?

A year ago now, I've opened a discussion in #37064 that proposes unifying the components that are used in the block inspector and global styles. The idea is that we'd have components that look like this:

```
function SomePanel( { value, onChange, inheritedValue } ) {

}
```

That component would render a panel used to edit a given style (say "typography" or "colors"), the `value` has a well specified prop (that is mapped over the "style" object of a block or a global style "node" ). The idea is that we could use that exact same panel in both places, the only difference is a wrapper that map the "value" to the right place:
 - block attributes for the block inspector
 - theme.json (global styles) for the global styles UI

Recently we realized this vision for the TypographyPanel and the current PR does the exact same thing for the DimensionsPanel and uses the new UI component for both inspector controls and global styles.

**To do**

 - [x] Check the extra layout panel in the block inspector.
 - [x] Find a way to ensure the visualizers work as intended in the block inspector.
 - [x] Test properly that there's no unexpected differences in behavior between this PR and trunk.

## Testing Instructions

 - you can test the "dimensions" panels in global styles and ensure that they still work as expected. 
 - you can also test the "dimensions" panels in the block inspector.

### CODE REVIEW

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/global-styles/dimensions-panel.js]:
```suggestion
	const resetMarginValue = () => setMarginValues( undefined );
```

Small copy and paste typo?

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/global-styles/dimensions-panel.js:210]: It appears that `useHasSpacingPresets` is only returning true for displaying the spacing presets control when in the Global Styles, not when styling individual blocks.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for the hard work on this @youknowriad. It's great to see so much of the boilerplate code being removed here. 👍 

So far, with the noted exception of the spacing visualizers, it is testing pretty well. The only issue that came up in a quick smoke test was that the spacing presets control isn't shown when editing individual blocks as it does on trunk.

| Before | After |
|---|---|
| Screenshot 2023-02-15 at 10 52 30 am | Screenshot 2023-02-15 at 11 06 09 am |



I'm happy to keep reviewing as this PR evolves though I'll be AFK until later next week.

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (youknowriad) [reply packages/block-editor/src/components/global-styles/dimensions-panel.js:210]: Should be fixed, good catch

File: packages/block-editor/src/hooks/utils.js
Comment (andrewserong) [inline packages/block-editor/src/hooks/utils.js:198]: Is it worth adding a doc block to this function? From reading through the changes, I wasn't sure what the intended scope is for settings included here — it looks like it's currently typography + spacing + layout, but not colors or position, for example. I imagine we'll continue to add to this function as each panel is extracted / consolidated, with the eventual goal for it to include all block settings?

Comment (youknowriad) [reply packages/block-editor/src/hooks/utils.js:198]: Yes, that's the goal I'm also thinking of making its result a locker prop in BlockEditContext, that way it's always available. Maybe I'll do the same for the "value" prop (the style extracted from all attributes) but I'm not 100% decided at this point.

**Comment by youknowriad** [comment]
I think this PR should be ready for final testing :)

**Review by andrewserong — COMMENTED** [review]
This is looking pretty close to me (and so good to see how this results in ~750 lines being removed). I ran into a couple of bugs in testing:

### Global styles

❌ contentSize (doesn't update in the editor canvas and doesn't appear to save)
❌ wideSize (doesn't update in the editor canvas and doesn't appear to save)
✅ padding
✅ margin
✅ block spacing
✅ minimum height

### Individual block

✅ padding
❌ margin (resetting individual value causes a block error)
✅ block spacing
✅ minimum height
✅ child layout controls (fit, fill, width)
✅ injected controls like min height of cover block, aspect ratio in post featured image
✅ visualizers for padding and margin

### Screengrabs

| contentSize / wideSize not updating / saving | resetting individual margin control on Group block causes block error |
| --- | --- |
|  |  |

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/dimensions-panel.js]: This appears to be the line that's throwing an error when resetting an individual margin value at the block level with the Group block. Possibly a related note here is that the Group block's opt in for margin only includes the top/bottom sides: `"margin": [ "top", "bottom" ]`.

image

**Comment by youknowriad** [comment]
Thanks for the great testing @andrewserong I've fixed these issues.

I'm not 100% satisfied with the fix for the layout properties in Global styles but I think it's a symptom of the fact that "layout" is actually stored in "settings" rather than "styles" in the global styles config. I'm not changing that today :) 

Also it could be improved a bit if we have a separate "LayoutPanel" like we do in the block inspector. Actually, in terms of UX we should try to unify a bit later, in inspector layout is separate from the dimensions panel but it's included there in global styles.

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (andrewserong) [reply packages/block-editor/src/components/global-styles/dimensions-panel.js]: Looks like this line was fixed in https://github.com/WordPress/gutenberg/pull/48070/commits/d055ea5dbf5553a5d6daa72b5ca9c440ba8231c1 and then possibly accidentally reverted in https://github.com/WordPress/gutenberg/pull/48070/commits/e46b52edf191726d80d1da9b97d28f5acab7be92? Resetting margin is currently causing the block error, but using the optional chaining fix from https://github.com/WordPress/gutenberg/pull/48070/commits/d055ea5dbf5553a5d6daa72b5ca9c440ba8231c1  fixes it for me 🙂

**Comment by andrewserong** [comment]
Thanks for updating @youknowriad, the Layout controls are working now. I see what you mean — there doesn't appear to be an easy way to workaround the fact of the difference between the global styles panel and the panel in the block editor when it comes to layout controls. I suppose another potential way to do it might be to inject the Layout controls into the DimensionsPanel's slot via inspector controls? However if we did that, then they'd wind up at the bottom of the list. I think the `includeLayoutControls` flag for now sounds like a good approach 👍

> Also it could be improved a bit if we have a separate "LayoutPanel" like we do in the block inspector. Actually, in terms of UX we should try to unify a bit later, in inspector layout is separate from the dimensions panel but it's included there in global styles.

On the topic of consolidating layout and dimensions controls, though, I believe both @tellthemachines and @SaxonF  have been advocating for the idea of working out how to combine them. I think the current status is that the Layout (and together with Dimensions) controls need some more design work to figure out how they could all sit nicely together, but it'd be a good thing to figure out in the longer-term as they're all very related.

Related issues:

* https://github.com/WordPress/gutenberg/issues/47902
* https://github.com/WordPress/gutenberg/issues/42385

Getting back to this PR, I think the last commit might have reverted the fix for the margin individual value resetting? Other than that, this appears to be testing really well. I'm at the end of my week now, but can give this a final review again on Monday.

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (youknowriad) [reply packages/block-editor/src/components/global-styles/dimensions-panel.js]: 🤕

File: packages/block-editor/src/components/global-styles/dimensions-panel.js
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/dimensions-panel.js]: It looks like this one should be `childLayout` as `layout` appears to be an SVG element. Would it be worth importing `layout` as `layoutIcon` so that it's harder to accidentally reuse `layout`?

**Review by andrewserong — COMMENTED** [review]
Thanks again for all the follow-up, here, it's looking very close to me. All the previously flagged issues have been resolved, the one issue I encountered in re-testing is that the child layout Width controls don't appear to be able to hidden once revealed, with this PR applied. When I go to reset the Width control after adding it, it resets the value, but doesn't hide it.

| Trunk | This PR |
| --- | --- |
|  |  |

Oh, actually, I see from inspecting the block markup, it looks like an SVG element is accidentally being injected into the `style.layout` attributes when it's being reset:

image

Looks like a one-liner to fix it up, I've left a comment 🙂

**Comment by youknowriad** [comment]
Thanks for the deep testing @andrewserong it should be fixed now

File: packages/block-editor/src/components/child-layout-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/child-layout-control/index.js:50]: Should we add the missing dependencies here or remove the `[]` arg to satisfy the linter?

File: packages/block-editor/src/components/child-layout-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/child-layout-control/index.js]: Nit: While we're relocating this code, should we avoid the nested `if` statements?

**Review by andrewserong — APPROVED** [review]
This is testing really well for me now, I haven't managed to break it again 😄

Thanks again for all the follow-up here, I think this is in good shape to land now, and I think the trade-offs made (e.g. layout controls in dimensions) make good sense, and we'll also be able to keep tweaking code quality things in follow-ups. One of the things I liked about the shape of the `DimensionsPanel` code in the site editor, is that we'd managed to further extract some of the inline definitions of things like `setPaddingValue` into hooks like `usePaddingProps`, so it might be worth pursuing that sort of approach again in the future, if we find that the `DimensionsPanel` becomes hard to read in its current form. But that's not at all a blocker, and just an idle thought from reading through the code again.

This LGTM now, so giving it the ✅ — do feel free to wait for another approving review if you're concerned about whether we've covered everything, but I'm feeling pretty confident about this change now after the last few rounds of testing and poking it 😄

Thanks again for all the continued work consolidating these components! ✨

**Comment by aaronrobertshaw** [comment]
> I think this is in good shape to land now,

FYI, I'm currently halfway through testing and reviewing this PR as well. There are a couple of small things we should probably tweak code-wise before merging. I'll complete the review shortly.

File: packages/block-editor/src/components/global-styles/hooks.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/global-styles/hooks.js]: Aren't these already handled in the `forEach` above?
```suggestion
```

File: packages/block-editor/src/components/global-styles/hooks.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/global-styles/hooks.js]: Should this be consolidated with the previous loop through `padding`, `margin`, and `blockGap`?

**Review by aaronrobertshaw — COMMENTED** [review]
Appreciate the hard work iterating on this one 🙇 

As Andrew noted, this is now testing really well. 

#### Global styles
✅ Padding & margin
✅ Block gap
✅ Min height
✅ ContentSize & wideSize
✅ Typography panel

#### Individual blocks
✅ Padding & margin
✅ Block gap
✅ Visualizers
✅ Min height
✅ Child layouts
✅ Non-block-support controls e.g.  aspect ratio in Post Feature Image
✅ Typography panel

I've left a few inline comments for minor things we might be able to clean up. 

While reviewing the files in this PR I also noticed a typo bug ([`typography.fuild`](https://github.com/WordPress/gutenberg/blob/8e0ffbddecec6daaacb6e8c86c7c20b0992061b9/packages/block-editor/src/components/global-styles/hooks.js#L58)) introduced in https://github.com/WordPress/gutenberg/pull/47356. I'm happy to throw up a quick fix or approve it as part of this PR if you'd prefer.

**Comment by aaronrobertshaw** [comment]
> One of the things I liked about the shape of the DimensionsPanel code in the site editor, is that we'd managed to further extract some of the inline definitions of things like setPaddingValue into hooks

This sounds like a good follow-up when bandwidth permits. I had similar thoughts reading through that section of the code.

**Comment by youknowriad** [comment]
> One of the things I liked about the shape of the DimensionsPanel code in the site editor, is that we'd managed to further extract some of the inline definitions of things like setPaddingValue into hooks

I saw this pattern :), I wondered whether I should apply it here, I'm hesitant and also not totally convinced.  I was wondering if sub components is better here than several hooks. That said I consider this as "implementation details" of the component itself and since it has no impact on the API, so I don't really care too much.

File: packages/block-editor/src/components/child-layout-control/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/child-layout-control/index.js:50]: Adding the missing dependencies here is actually a change in behavior. While it's probably the right thing to do, I prefer not to do it in this PR because I don't really know this code and might introduce breakage.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/child-layout-control/index.js:50]: Good call. @andrewserong and @tellthemachines might be familiar enough to give a 👍 or 👎 in a follow-up.

**Comment by youknowriad** [comment]
I think I've addressed the suggestion here. Thanks all for the reviews, I really appreciate it. 

Let's see if we can continue with other panels.

**Comment by andrewserong** [comment]
Nice work landing this one!

> I wondered whether I should apply it here, I'm hesitant and also not totally convinced. I was wondering if sub components is better here than several hooks. That said I consider this as "implementation details" of the component itself and since it has no impact on the API, so I don't really care too much.

I agree, it's definitely just the "implementation details" side of things, and sub-components would be a good thing to explore and neater than a proliferation of hooks 👍. From my perspective, as these components get bigger, it'd be good to try to tidy them up a bit where we can to improve readability, but definitely not a blocker for anything 🙂

**Comment by aaronrobertshaw** [comment]
I put up a fix for the typo breaking the fluid typography setting in https://github.com/WordPress/gutenberg/pull/48605.

**Comment by aaronrobertshaw** [comment]
This PR introduced a small regression to the resetting of dimensions controls that have only axial values (horizontal & vertical) e.g. the Button block. 

With this PR, the `onDeselect` callbacks call `setPadding` etc which in turn call [`filterValuesBySides`](https://github.com/WordPress/gutenberg/pull/48070/files#diff-f8637021cd834da7a945db08831dd15b2ddcee2768b735c94966e7fb4f34ce9aR104-R125) which assumes there's a values object.

A fix is up in https://github.com/WordPress/gutenberg/pull/50654

**Pull request merged**
